File: SampleAppObjectiveC/SkylinkSample/TEMRichVideoView.m
//
//  TEMRichVideoView.m
//  SampleApp
//
//  Created by macbookpro on 07/05/2015.
//  Copyright (c) 2015 Temasys Communications. All rights reserved.
//

#import "TEMRichVideoView.h"

#import "TEMProgressView.h"

#define LABEL_HEIGHT 21

@interface TEMRichVideoView ()

@property(nonatomic, weak) TEMProgressView* progressView;

@end

@implementation TEMRichVideoView

static void init(TEMRichVideoView *self, UIView *renderView) {
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, self.bounds.size.width, LABEL_HEIGHT)];
    label.textColor = [UIColor blueColor];
    label.textAlignment = NSTextAlignmentLeft;
    label.numberOfLines = 1;
    label.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    label.translatesAutoresizingMaskIntoConstraints = YES;
    [self insertSubview:label aboveSubview:renderView];
    self.lblTitle = label;
}

- (id)initWithFrame:(CGRect)frame videoView:(UIView*)videoView {
    self = [super initWithFrame:frame videoView:videoView];
    if (self) {
        init(self, videoView);
        _isEnabled = TRUE;
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    UIView *renderSurface = [self getRenderSurface];
    self.lblTitle.frame = CGRectMake(renderSurface.frame.origin.x, renderSurface.frame.origin.y, renderSurface.frame.size.width, LABEL_HEIGHT);
    self.progressView.frame = CGRectMake(self.lblTitle.frame.origin.x, CGRectGetMaxY(self.lblTitle.frame), self.lblTitle.frame.size.width, self.lblTitle.frame.size.height);
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)addProgressView {
    // Listen to the 'SKYLINKFileProgress' Notification.
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(gotProgress:) name:@"SKYLINKFileProgress" object:nil];
    TEMProgressView *progressView = [[TEMProgressView alloc] initWithFrame:CGRectMake(self.lblTitle.frame.origin.x, CGRectGetMaxY(self.lblTitle.frame), self.lblTitle.frame.size.width, self.lblTitle.frame.size.height)];
    [self insertSubview:progressView aboveSubview:self.lblTitle];
    self.progressView = progressView;
}

- (void)removeProgressView {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    [self.progressView removeFromSuperview];
    self.progressView = nil;
}

#pragma mark - Properties

- (void)setIsEnabled:(BOOL)isEnabled {
    if (_isEnabled != isEnabled) {
        _isEnabled = isEnabled;
        if (isEnabled) {
            [self getTouchSurface].alpha = 1.0;
            [self getTouchSurface].backgroundColor = [UIColor clearColor];
        } else {
            [self getTouchSurface].alpha = 0.5;
            [self getTouchSurface].backgroundColor = [UIColor lightGrayColor];
        }
    }
}

- (void)setIsRemote:(BOOL)isRemote {
    if (_isRemote != isRemote) {
        _isRemote = isRemote;
        if (isRemote)
            [self getRenderSurface].transform = CGAffineTransformMakeScale(-1, 1);
    }
}

- (NSString*)title {
    return self.lblTitle.text;
}

- (void)setTitle:(NSString *)_title {
    self.lblTitle.text = _title;
}

#pragma mark - Private

/**
 @discussion Update progress bar upon getting 'SKYLINKFileProgress' Notification.
 */
- (void)gotProgress:(NSNotification*)notification {
    if ([self.ID caseInsensitiveCompare:[notification.userInfo objectForKey:@"peerId"]] == NSOrderedSame)
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.progressView setPercentage:[[notification.userInfo objectForKey:@"percentage"] floatValue]];
        });
}

@end

File: SampleAppSwift/SampleAppSwift/TEMRichVideoView.swift
//
//  TEMRichVideoView.swift
//  SampleAppSwift
//
//  Created by macbookpro on 08/05/2015.
//  Copyright (c) 2015 Temasys Communications. All rights reserved.
//

import UIKit

class TEMRichVideoView: TEMVideoView {
    
    var ID: String = ""
    
    weak var lblTitle: UILabel?
    
    private weak var progressView: TEMProgressView?
    
    private var enabled: Bool = true
    private var remote: Bool = true
    
    static func richInitialize(videoView: TEMRichVideoView, renderView: UIView) {
        let selfSize: CGSize = videoView.bounds.size;
        var label: UILabel = UILabel(frame: CGRectMake(0, 0, selfSize.width, 21))
        label.textColor = UIColor.blueColor()
        label.textAlignment = NSTextAlignment.Left;
        label.numberOfLines = 1;
        label.autoresizingMask = UIViewAutoresizing.FlexibleWidth;
        label.setTranslatesAutoresizingMaskIntoConstraints(true)
        videoView.insertSubview(label, aboveSubview: renderView)
        videoView.lblTitle = label;
    }
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override init(frame: CGRect, videoView: UIView) {
        super.init(frame: frame, videoView: videoView)
        TEMRichVideoView.richInitialize(self, renderView: videoView)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        var renderSurface: UIView = self.getRenderSurface()
        self.lblTitle?.frame = CGRectMake(renderSurface.frame.origin.x, renderSurface.frame.origin.y, renderSurface.frame.size.width, 21)
        if (self.progressView != nil) {
            let lblTitleFrame: CGRect! = self.lblTitle?.frame
            self.progressView?.frame = CGRectMake(lblTitleFrame.origin.x, CGRectGetMaxY(lblTitleFrame), lblTitleFrame.size.width, lblTitleFrame.size.height)
        }
    }
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    // MARK: - public methods
    
    func addProgressView() {
        // Listen to the 'SKYLINKFileProgress' Notification.
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "gotProgress", name: "SKYLINKFileProgress", object: nil)
        let lblTitleFrame: CGRect! = self.lblTitle?.frame
        var progressView: TEMProgressView = TEMProgressView(frame: CGRectMake(lblTitleFrame.origin.x, CGRectGetMaxY(lblTitleFrame), lblTitleFrame.size.width, lblTitleFrame.size.height))
        self.insertSubview(progressView, aboveSubview: self.lblTitle!)
        self.progressView = progressView;
    }

    func removeProgressView() {
        NSNotificationCenter.defaultCenter().removeObserver(self)
        self.progressView?.removeFromSuperview()
        self.progressView = nil
    }
    
    // MARK: - properties
    
    func getTitle() -> String? {
        return self.lblTitle?.text
    }
    
    func setTitle(text: String) {
        self.lblTitle?.text = text
    }
    
    func isEnabled() -> Bool {
        return self.enabled
    }
    
    func setEnabled(enabled: Bool) {
        self.enabled = enabled
        if (self.enabled != enabled) {
            self.enabled = enabled;
            if (enabled) {
                self.getTouchSurface().alpha = 1.0
                self.getTouchSurface().backgroundColor = UIColor.clearColor()
            } else {
                self.getTouchSurface().alpha = 0.5
                self.getTouchSurface().backgroundColor = UIColor.lightGrayColor()
            }
        }
    }
    
    func isRemote() -> Bool {
        return self.remote
    }
    
    func setRemote(remote: Bool) {
        if (self.remote != remote) {
            self.remote = remote;
            if (remote) {
                self.getRenderSurface().transform = CGAffineTransformMakeScale(-1, 1)
            }
        }
    }
    
    // MARK: - private methods
    
    /**
    @discussion Update progress bar upon getting 'SKYLINKFileProgress' Notification.
    */
    func gotProgress(notification: NSNotification) {
        if let userInfo = notification.userInfo {
            let comparisonResult: NSComparisonResult = self.ID.caseInsensitiveCompare(userInfo["peerId"] as! String)
            if (comparisonResult == NSComparisonResult.OrderedSame) {
                dispatch_async(dispatch_get_main_queue(), { () -> Void in
                    self.progressView?.setPercentage(userInfo["percentage"] as! CGFloat)
                })
            }
        }
    }
    
}
